apps/api/src/sub_refresh_job.rs

```diff
@@ -0,0 +1,145 @@
+	transactions::{TransactionError, TransactionResult, TransactionSession},
+};
+
+pub fn sub_age_days(periods: &[SubscriptionPeriod]) -> isize {
+	sub_age(periods).0
+}
+
+pub fn sub_age(periods: &[SubscriptionPeriod]) -> (isize, isize) {
+	periods
+		.iter()
+		.map(|p| {
+			let diff = date_component::date_component::calculate(&p.start, &p.end);
+			(diff.interval_days, diff.month)
+		})
+		.fold((0, 0), |(days, month), diff| (days + diff.0, month + diff.1))
+}
+
+/// Grants entitlements for a subscription.
+pub async fn refresh_entitlements(
+	tx: &mut TransactionSession<'_, ApiError>,
+	subscription_id: &SubscriptionId,
+	benefits: &[SubscriptionBenefit],
+) -> TransactionResult<(), ApiError> {
+	// first revoke all entitlements
+	revoke_entitlements(tx, subscription_id).await?;
+
+	// load all periods
+	let periods = tx
+		.find(
+			filter::filter! {
+				SubscriptionPeriod {
+					#[query(serde)]
+					subscription_id: subscription_id,
+				}
+			},
+			None,
+		)
+		.await?;
+
+	let now = chrono::Utc::now();
+	let active = periods.iter().any(|p| p.start < now && p.end > now);
+
+	if active {
+		let mut edges = vec![];
+
+		// always insert the edge from the user to the subscription
+		edges.push(EntitlementEdgeId {
+			from: EntitlementEdgeKind::User {
+				user_id: subscription_id.user_id,
+			},
+			to: EntitlementEdgeKind::Subscription {
+				subscription_id: subscription_id.clone(),
+			},
+			managed_by: Some(EntitlementEdgeManagedBy::Subscription {
+				subscription_id: subscription_id.clone(),
+			}),
+		});
+
+		let (age_days, age_months) = sub_age(&periods);
+
+		for benefit in benefits {
+			let is_fulfilled = match &benefit.condition {
+				SubscriptionBenefitCondition::Duration(DurationUnit::Days(d)) => age_days >= (*d as isize),
+				SubscriptionBenefitCondition::Duration(DurationUnit::Months(m)) => age_months >= (*m as isize),
+				SubscriptionBenefitCondition::TimePeriod(tp) => {
+					periods.iter().any(|p| p.start <= tp.start && p.end >= tp.end)
+				}
+			};
+
+			if is_fulfilled {
+				edges.push(EntitlementEdgeId {
+					from: EntitlementEdgeKind::Subscription {
+						subscription_id: subscription_id.clone(),
+					},
+					to: EntitlementEdgeKind::SubscriptionBenefit {
+						subscription_benefit_id: benefit.id,
+					},
+					managed_by: Some(EntitlementEdgeManagedBy::Subscription {
+						subscription_id: subscription_id.clone(),
+					}),
+				});
+			}
+		}
+
+		tx.insert_many(edges.into_iter().map(|id| EntitlementEdge { id }), None)
+			.await?;
+	}
+
+	Ok(())
+}
+
+/// Grants entitlements for a subscription.
+pub async fn revoke_entitlements(
+	tx: &mut TransactionSession<'_, ApiError>,
+	subscription_id: &SubscriptionId,
+) -> TransactionResult<(), ApiError> {
+	tx.delete(
+		filter::filter! {
+			EntitlementEdge {
+				#[query(rename = "_id", flatten)]
+				id: EntitlementEdgeId {
+					#[query(serde)]
+					managed_by: Some(EntitlementEdgeManagedBy::Subscription { subscription_id: subscription_id.clone() }),
```

Comment:
I dont think this is correct, in theory there should only be one edge where this is set.


User -> Subscription 

This edge is the only edge where the EntitlementEdgeManagedBy will be set.

This is because this is the only edge that is managed by the subscription that is not a direct decedent of the subscription.